File: 463-island-perimeter/463-island-perimeter.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        ans = [0]
        visited = set()
        dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        
        def inBounds(row, col):
            return row in range(ROWS) and col in range(COLS)
        
        def isLand(row, col):
            return inBounds(row, col) and grid[row][col] == 1
        
        def dfs(row, col, count):
            visited.add((row, col))
            for rowDiff, colDiff in dirs:
                newRow, newCol = row + rowDiff, col + colDiff
                if isLand(newRow, newCol) and (newRow, newCol) not in visited:
                    dfs(newRow, newCol, count)
                elif not isLand(newRow, newCol):
                    count[0] += 1
            return
                    
        for row in range(ROWS):
            for col in range(COLS):
                if isLand(row, col):
                    dfs(row, col, ans)
                    return ans[0]

                    

File: 463-island-perimeter/463-island-perimeter.rb
# @param {Integer[][]} grid
# @return {Integer}

def in_bounds(row, col, grid)
    return row >= 0 && row < grid.length && col >= 0 && col < grid[0].length
end

def is_land(row, col, grid)
    return in_bounds(row, col, grid) && grid[row][col] == 1
end

def dfs(row, col, grid, visited, count)
    visited.add([row, col])
    dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    dirs.each do |row_diff, col_diff|
        new_row, new_col = row + row_diff, col + col_diff
        if is_land(new_row, new_col, grid) && !visited.include?([new_row, new_col])
            dfs(new_row, new_col, grid, visited, count)
        elsif !is_land(new_row, new_col, grid)
            count[0] += 1
        end
    end
end

def island_perimeter(grid)
    ans = [0]
    visited = Set.new
    
    (0...grid.length).each do |row|
        (0...grid[row].length).each do |col|
            if is_land(row, col, grid)
                dfs(row, col, grid, visited, ans)
                return ans[0]
            end
        end
    end
end